src/app/d2-loadout-builder/generated-sets/GeneratedSets.tsx

```diff
@@ -43,13 +47,42 @@ export default class GeneratedSets extends React.Component<Props, State> {
 
   componentWillReceiveProps(props: Props) {
     if (props.processedSets !== this.props.processedSets) {
-      this.setState({ minimumPower: 0 });
+      this.setState({ minimumPower: 0, shownSets: 10 });
     }
   }
 
+  showMore = () => {
+    this.setState({ shownSets: this.state.shownSets + 10 });
+  };
+
+  setSelectedTier = (element) => {
+    this.setState({ shownSets: 10, selectedTier: element.target.value });
+  };
+
+  setMinimumPower = (element) => {
+    this.setState({ shownSets: 10, minimumPower: parseInt(element.target.value, 10) });
+  };
+
+  toggleLockedItem = (lockedItem: LockedItemType) => {
+    let newLockedItems: LockedItemType[] = [];
+    const bucket = (lockedItem.item as D2Item).bucket;
+    if (this.props.lockedMap[bucket.hash] && this.props.lockedMap[bucket.hash][0].type !== 'item') {
+      newLockedItems = this.props.lockedMap[bucket.hash];
+    }
+
+    const existingIndex = newLockedItems.findIndex((existing) => existing.item === lockedItem.item);
+    if (existingIndex > -1) {
+      newLockedItems.splice(existingIndex, 1);
+    } else {
+      newLockedItems.push(lockedItem);
+    }
```

Comment:
It's better to make a new array and set that, than to modify the existing array in place. If you replace it, we can play fun tricks with reference equality to do fast updates.

Reply:
like so? https://github.com/DestinyItemManager/DIM/pull/3193/commits/a0a2821fd844b1526271452e3c00b64f076c0154